### ISSUE TITLE
[VSCode][Protocols] Inconsistent errors when using protocols with immutable data containers

### ISSUE THREAD

**Issue by MadhuVK** [microsoft/pyright#6226]
**Describe the bug**

1. **Behavior**: Able to assign a NamedTuple to a protocol with a read/write property; unable to modify the property.
   **Expected**: Expected the error earlier, when assigning to variable typed with the Protocol.
   
1. **Behavior**: Able to assign a frozen dataclass to a protocol with a read/write property; unable to modify the property.
   **Expected**: Expected the error earlier, when assigning to variable typed with the Protocol.
   
1. **Behavior**: Able to assign a frozen dataclass to a protocol with a read/write property using variable annotation defined in PEP-0526; unable to modify the property.
   **Expected**: Expected the error earlier, when assigning to variable typed with the Protocol.

**Note**: This appears to be working as expected with NamedTuples and read/write properties defined a la PEP-0526 (the error is at the assignment).

**Code or Screenshots**
![screen](https://github.com/microsoft/pyright/assets/1362193/b57c2770-202a-4664-80b8-1a58716f0d46)

```python
#!/usr/bin/env python3
from dataclasses import dataclass
from typing import NamedTuple, Protocol

class DataNamed(NamedTuple):
    name: str = "Name"

@dataclass(frozen=False)
class DataClass:
    name: str = "Name"

@dataclass(frozen=True)
class DataClassFrozen:
    name: str = "Name"

# Protocol: RW defined via property annotations (PEP-0526)
class DataPropertyRW(Protocol):
    @property
    def name(self) -> str: ...
    @name.setter
    def name(self, n: str) -> None: ...

rwp_named: DataPropertyRW = DataNamed()             # Expected an error since DataNamed is RO
rwp_named.name = "NewName"                          # Error is good, but would have preferred line in line above

rwp_frozenclass: DataPropertyRW = DataClassFrozen() # Expected an error because a frozen class
rwp_frozenclass.name = "NewName"                    # Error is good, but would have preferred in the line above

rwp_class: DataPropertyRW = DataClass()             # Normal
rwp_class.name = "NewName"                          # Normal

# Protocol: RW Property defined via variable annotation (PEP-0526)
class VariableAnnotatedRW(Protocol):
    name: str

rwva_named: VariableAnnotatedRW = DataNamed()             # Good error
rwva_named.name = "NewName"

rwva_frozenclass: VariableAnnotatedRW = DataClassFrozen() # Expected an error because a frozen class
rwva_frozenclass.name = "NewName"                         # Error is good, but would have preferred in the line above

rwva_class: VariableAnnotatedRW = DataClass()             # Normal
rwva_class.name = "NewName"                               # Normal
```

**VS Code extension or command-line**
Python 3.10.13
Python VSCode Version v2023.18.0

**Comment by erictraut** [comment]
Thanks for the bug report. I appreciate the great code sample and thoroughness of your analysis.

You are correct that the protocol matching code didn't contain any special casing for named tuples or frozen dataclasses, so it was treating these as mutable. I've addressed the short coming, and this will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.333, which I just published. It will also be included in a future release of pylance.

**Comment by MadhuVK** [comment]
Great, thanks for the quick ack and turnaround here!
